lessons-learned-from-some-of-the-best-ruby-codebases-part-3.markdown

```diff
@@ -0,0 +1,470 @@
+
+One thing I'd quickly like to talk about though is this:
+
+```Ruby
+module Adamantium
+  class Freezer
+    class Deep < self  # <- Wat?
+      # snip
+    end
+  end
+end
+```
+
+`Deep` inherits from `self`. And what is `self` here?
+
+```Ruby
+class Omg; puts self; class Bar < self; end; end
+# => Omg
+```
+
+So
+
+```Ruby
+class Deep < self
+```
+
+is actually a fancy way of writing
+
+```Ruby
+class Deep < Freezer
+```
+
+But why are we doing this here instead of being explicit? It's making our intention explicit without duplicating names across our module. And even if you rename `Freezer` to something else, this line doesn't change, which makes it easier to refactor and less error-prone to subtle bugs.
+
+Is it a huge deal? No. It isn't. But in a larger codebase small things like can make the difference between "bareley maintainable" and "updating functionality is a walk in the park".
+
+Before we wrap this up here let's look at one last thing:
+
+We saw how the default strategy is applied. What about that the `flat` strategy?
+
+A quick reminder to how applying and using the `flat` strategy looked like:
+
+```Ruby
+class FlatExample
+  include Adamantium::Flat
+  #
+  # object = Example.new
+  # object.random => ["abcdef"]
+  # object.random => ["abcdef"]
+  # object.random.frozen? => true
+  # object.random[0].frozen? => false
+  #
+  def random
+    [SecureRandom.hex(6)]
+  end
+  memoize :random
+end
+```
+
+As you can see, you just include the `Flat` submodule of `Adamantium`.
+
+How is this implemented?
+
+Let's look at the `included` callback:
+
+```Ruby
+module Adamantium
+  module Flat
+    def freezer
+      Freezer::Flat
+    end
+
+    def self.included(descendant)
+      descendant.instance_exec(self) do |mod|
+        include Adamantium
+        extend mod
+      end
+    end
+  end
+end
+```
+
+That's pretty cool. Let me walk you through it:
+
+```Ruby
+descendant.instance_exec(self) do |mod|
+  # snip
+end
+```
+
+We're calling `instance_exec` on the descendant and pass `self` in. What's `self` now?
+It's the module constant, so...`Adamantium::Flat`.
+We then include `Adamantium`. That is something we already covered above in this article. So bussiness as usual.
+But then we extend `mod`, so `Adamantium::Flat`. Why? Because this basically overwrites the `freezer` methods that we imported in our class by doing
+
+```Ruby
+include Adamantium
+```
+
+so that this now returns the `flat` strategy
+
+```Ruby
+def freezer
+  Freezer::Flat
+end
+```
+
+not the `deep` strategy
+
+```Ruby
+def freezer
+  Freezer::Deep
+end
+```
+
+anymore.
+
+So long story short, this is just a very advanced configuration mechanism. Granted, it might seem a little complicated but at the same time it completely separates the modules from each other making it easier to update one without updating the other.
+
```

Comment:
This post = awesome.

Since it contains a lot of useful information I think a summary along the lines of "what have we learned so far" would be great here.

Also, I think a couple of scenarios where freezing (or preventing abstract types for being instantiated) saves the day would be a good eye-catcher.